<a href="https://colab.research.google.com/github/himu0023/GEN-AI/blob/main/DL/ANN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Importing the dataset
df = pd.read_csv("/content/drive/MyDrive/ColabNotebooks/GenAi/DEEP LEARNING/Churn_Modelling.csv")
X = df.iloc[:,3:13].values
y = df.iloc[:,13].values

In [38]:
df.dtypes

RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                float64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object

In [4]:
X

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]],
      shape=(10000, 10), dtype=object)

In [5]:
y

array([1, 0, 1, ..., 1, 1, 0], shape=(10000,))

In [6]:
# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labellencoder_X_1 = LabelEncoder()
X[:, 1] = labellencoder_X_1.fit_transform(X[:, 1])
labellencoder_X_2 = LabelEncoder()
X[:, 2] = labellencoder_X_2.fit_transform(X[:, 2])
X

array([[619, 0, 0, ..., 1, 1, 101348.88],
       [608, 2, 0, ..., 0, 1, 112542.58],
       [502, 0, 0, ..., 1, 0, 113931.57],
       ...,
       [709, 0, 0, ..., 0, 1, 42085.58],
       [772, 1, 1, ..., 1, 0, 92888.52],
       [792, 0, 0, ..., 1, 0, 38190.78]], shape=(10000, 10), dtype=object)

In [7]:
# Splitting the dataset into Traning and Test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2, random_state = 42)

In [8]:
len(df)

10000

In [9]:
len(X_test)

2000

In [10]:
# Feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [11]:
# Decision Tree
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()
clf.fit(X_train, y_train)
y_preds = clf.predict(X_test)

# Making the confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_preds)
print(cm)

# Accuracy
from sklearn.metrics import accuracy_score
score = accuracy_score(y_test, y_preds)
print(f"Accuracy of Decision Tree: {score*100:.2f}%")

[[1330  277]
 [ 194  199]]
Accuracy of Decision Tree: 76.45%


In [12]:
# Random forest Classifier
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=200)
clf.fit(X_train, y_train)
y_preds2 = clf.predict(X_test)

# Making the confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_preds)
print(cm)

# Accuracy
from sklearn.metrics import accuracy_score
score = accuracy_score(y_test, y_preds2)
print(f"Accuracy of Random Forest: {score*100:.2f}%")

[[1330  277]
 [ 194  199]]
Accuracy of Random Forest: 86.80%


## ANN - Let's get started !!!

In [13]:
# !pip install keras
# !pip install tensorflow

In [14]:
# Importing the Keras Libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

In [15]:
# Initilising the ANN
classifier = Sequential()

# units : It's an art, comes by experience, input+output/2
# kernel_initializer: how out weights are updated
# relu activation function
# Input_dim: input layer and the firest hidden layer

# Adding the input layer and the first hidden layer
classifier.add(Dense(units = 6, kernel_initializer= 'uniform', activation = 'relu', input_dim = 10))

# Adding the second hidden layer
classifier.add(Dense(units = 6, kernel_initializer = "uniform", activation = 'relu'))

# Adding the third hidden layer
classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))

# Adding the output layer
classifier.add(Dense(units = 1, kernel_initializer= 'uniform', activation = 'sigmoid')) # softmax

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [16]:
len(X_test)

2000

In [17]:
# Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics=['accuracy']) # categorical_crossentropy or sparse_categorical_crossentropy
y_train[0]

# Fitting the ANN to the Trainig set
classifier.fit(X_train, y_train, batch_size=10, epochs=50)

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8038 - loss: nan
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7979 - loss: nan
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7893 - loss: nan
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7915 - loss: nan
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7906 - loss: nan
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7920 - loss: nan
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7954 - loss: nan
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7906 - loss: nan
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7893 - loss: nan
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7949 - loss: nan
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7914 - loss: nan
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7947 - loss: nan
E

In [20]:
# Part 3 - Making prediction and evaluation the model
# Predicting the test result

y_pred_ann = classifier.predict(X_test)
y_pred_ann = (y_pred_ann > 0.5)

cm3 = confusion_matrix(y_test, y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step  
[[1607    0]
 [ 393    0]]


In [24]:
score3 = accuracy_score(y_test, y_pred_ann)
print(f"Accuracy of ANN: {score3*100:.2f}%")

Accuracy of ANN: 80.35%


## Accuracy from ANN: 80.35%

# Conclusion:
So, as we have already seen, by using 2 hidden layered ANN, we aren't able to get beyond 80% accuracy, of course we can get higher accuracy by increasing the epochs and doung some hyperparameter tuning, so on the other hand, the classical models can also be tuned further.

Bottom line, always start with classical machine learning models, and of we aren't able to increase the performance, then move towards Deep Learning techniques.

In [17]:
# Predicting a single new observation
"""Predict if the customer with the following information will leave the bank:
Geography: France
Credit Score: 600
Gender: Male
Age: 40
Tenure: 3
Balance: 60000
Number of Products: 2
Has Credit Card: Yes
Is Active Member: Yes
Estimated Salary: 50000"""

new_data = np.array([[600, 0, 1, 40, 3, 60000, 2, 1, 1, 50000]])  # 2D
new_data = sc.transform(new_data)
new_prediction = classifier.predict(new_data)
new_prediction = (new_prediction > 0.5)
new_prediction

# Making the confusion matrix
# from sklearn.metrics import confusion_matrix
# cm = confusion_matrix(y_test, y_preds)
# print(cm)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 95ms/step


array([[False]])

In [18]:
# !pip install scikeras


In [26]:
import numpy as np
import pandas as pd

# scikit-learn
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.model_selection import cross_val_score

# Keras / TensorFlow
from scikeras.wrappers import KerasClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# ---------------------------
# Dataset
# ---------------------------
# X = df.iloc[:, 3:13].values   # 10 input columns
# y = df.iloc[:, 13].values     # target column

# Encode target
le = LabelEncoder()
y = le.fit_transform(y)

# ---------------------------
# Preprocessing
# ---------------------------
categorical_features = [1, 2]   # Geography, Gender
numeric_features = [0, 3, 4, 5, 6, 7, 8, 9]

numeric_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="mean")),
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer([
    ("num", numeric_transformer, numeric_features),
    ("cat", categorical_transformer, categorical_features)
])

# ---------------------------
# Model builder
# ---------------------------
def build_classifier(meta):
    model = Sequential()
    model.add(Dense(units=6, activation="relu", input_dim=meta["n_features_in_"]))
    model.add(Dense(units=6, activation="relu"))
    model.add(Dense(units=1, activation="sigmoid"))
    model.compile(optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"])
    return model

classifier = KerasClassifier(
    model=build_classifier,
    epochs=10,
    batch_size=10,
    verbose=0
)

# ---------------------------
# Full pipeline
# ---------------------------
clf = Pipeline([
    ("preprocessor", preprocessor),
    ("nn", classifier)
])

# ---------------------------
# Verify preprocessing
# ---------------------------
Xt = preprocessor.fit_transform(X)
print("Any NaNs left?", np.isnan(Xt).any())   # must be False
print("Shape after preprocess:", Xt.shape)

# ---------------------------
# Cross-validation
# ---------------------------
accuracies = cross_val_score(clf, X, y, cv=10, n_jobs=-1, error_score="raise")

print("Mean accuracy:", accuracies.mean())
print("Std deviation:", accuracies.std())


Any NaNs left? False
Shape after preprocess: (10000, 14)
Mean accuracy: 0.8523
Std deviation: 0.011393419153177866


In [ ]:
# More robust version with additional checks
import numpy as np
import pandas as pd

# scikit-learn
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.model_selection import GridSearchCV

# Keras / TensorFlow
from scikeras.wrappers import KerasClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# ---------------------------
# Data preparation with checks
# ---------------------------
X = df.drop(['RowNumber', 'CustomerId', 'Surname', 'Exited'], axis=1)
y = df['Exited']

print("Data shapes:")
print(f"X shape: {X.shape}")
print(f"y shape: {y.shape}")

print("\nData types:")
print(X.dtypes)

print("\nMissing values:")
print(X.isnull().sum())

# ---------------------------
# Preprocessing
# ---------------------------
categorical_features = ['Geography', 'Gender']
numeric_features = ['CreditScore', 'Age', 'Tenure', 'Balance',
                   'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary']

# Use simpler imputers that are more robust
numeric_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),  # More robust than mean
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="constant", fill_value="missing")),  # More robust
    ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
])

preprocessor = ColumnTransformer([
    ("num", numeric_transformer, numeric_features),
    ("cat", categorical_transformer, categorical_features)
])

# ---------------------------
# Test preprocessing first
# ---------------------------
print("\nTesting preprocessing...")
X_transformed = preprocessor.fit_transform(X)
print(f"Transformed shape: {X_transformed.shape}")

# ---------------------------
# Model builder
# ---------------------------
def build_classifier(meta, optimizer="adam", units1=6, units2=6):
    n_features_in_ = meta["n_features_in_"]

    model = Sequential()
    model.add(Dense(units=units1, activation="relu", input_dim=n_features_in_))
    model.add(Dense(units=units2, activation="relu"))
    model.add(Dense(units=1, activation="sigmoid"))
    model.compile(optimizer=optimizer,
                  loss="binary_crossentropy",
                  metrics=["accuracy"])
    return model

# ---------------------------
# Start with simpler grid search
# ---------------------------
classifier = KerasClassifier(
    model=build_classifier,
    epochs=10,  # Start with fewer epochs for testing
    batch_size=32,
    verbose=0,
    optimizer="adam"
)

clf = Pipeline([
    ("preprocessor", preprocessor),
    ("nn", classifier)
])

# Start with a smaller grid for testing
simple_param_grid = {
    "nn__epochs": [10, 20],
    "nn__batch_size": [32, 64],
    "nn__optimizer": ["adam"],
    "nn__model__units1": [6],
    "nn__model__units2": [6]
}

print("\nStarting grid search with simple parameters...")
grid = GridSearchCV(estimator=clf,
                    param_grid=simple_param_grid,
                    cv=2,  # Fewer folds for testing
                    n_jobs=1,
                    scoring="accuracy",
                    verbose=1)

grid_result = grid.fit(X, y)

print("Best score:", grid_result.best_score_)
print("Best params:", grid_result.best_params_)

# If successful, run the full grid search
print("\nStarting full grid search...")
full_param_grid = {
    "nn__epochs": [50, 100],
    "nn__batch_size": [10, 25],
    "nn__optimizer": ["adam", "rmsprop"],
    "nn__model__units1": [6, 12],
    "nn__model__units2": [6, 12]
}

full_grid = GridSearchCV(estimator=clf,
                         param_grid=full_param_grid,
                         cv=3,
                         n_jobs=-1,  # Use all cores
                         scoring="accuracy",
                         verbose=1)

full_grid_result = full_grid.fit(X, y)

print("Full grid search - Best score:", full_grid_result.best_score_)
print("Full grid search - Best params:", full_grid_result.best_params_)

Data shapes:
X shape: (10000, 10)
y shape: (10000,)

Data types:
CreditScore          int64
Geography           object
Gender              object
Age                float64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
dtype: object

Missing values:
CreditScore          0
Geography            0
Gender              54
Age                300
Tenure               0
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
dtype: int64

Testing preprocessing...
Transformed shape: (10000, 14)

Starting grid search with simple parameters...
Fitting 2 folds for each of 4 candidates, totalling 8 fits


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Best score: 0.8454
Best params: {'nn__batch_size': 32, 'nn__epochs': 10, 'nn__model__units1': 6, 'nn__model__units2': 6, 'nn__optimizer': 'adam'}

Starting full grid search...
Fitting 3 folds for each of 32 candidates, totalling 96 fits


In [ ]:
#Improvising the ANN - Hyper parameter optimization, adding more hyperparameters
######################################
######################################
######################################
######################################
#TO DO TASK
######################################
######################################
######################################
######################################

from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
def build_classifier(loss, optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 10))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = optimizer, loss = loss, metrics = ['accuracy'])
    return classifier


classifier = KerasClassifier(build_fn = build_classifier)

parameters = {'batch_size': [10,20, 30, 50],
              'epochs': [10,50,100,200],
              'loss': ['binary_crossentropy', 'categorical_crossentropy'],
              'optimizer': ['adam', 'rmsprop']}

grid_search = GridSearchCV(estimator = classifier, param_grid = parameters, scoring = 'accuracy', n_jobs = -1, cv = 10)

grid_search = grid_search.fit(X_train, y_train)

In [ ]:
tuple1 = ([2,3,4],6,7)

In [ ]:
tuple1[0][1]=5

In [ ]:
tuple1

## Improvising the performance of ANN

1. Increasing the number of hidden layers
2. Increasing the number of epochs
3. Fine-tuning the parameters - optimizers, batch_size, epochs
4. Dropout regularization: Dropout(0.25)

Overfitting problem --> training accuracy increases, but test accuracy is low